### ISSUE TITLE
feature: custom address labels

### ISSUE THREAD

**Issue by totlsota** [wmitsuda/otterscan#245]
This is the number one aspect that keeps me checking other tools/scripts/etherscan. 
With thanks, otterscan is great.

**Comment by wmitsuda** [comment]
You mean, privately attribute names to certain addresses? Sounds good, I just wonder what would be a good UX for persisting this data.

The simplest way would be to store it in browser local storage, but I've heard some time ago Apple would start deleting local storage after some time.

Do you have any idea of a better place you would like this personal data to be stored?

**Comment by totlsota** [comment]
Yes, privately attribute names to addresses: my contracts, other contracts of note. Etherscan has 'tags' and you can add your own 'tags' that will show throughout the UI. For example, I've tagged both the EOA and contract in this example:

<img width="1309" alt="image" src="https://user-images.githubusercontent.com/3116579/163816907-569ad810-22fe-4edc-89a0-0371142f64cb.png">

You can see the default 'tag' from etherscan too [here](https://etherscan.io/address/0x00000000726422a6fECb4759b44D47E48Cf746aa).

<img width="695" alt="image" src="https://user-images.githubusercontent.com/3116579/163817006-faf7b9a4-00f6-4b03-83e6-e0d9d934774d.png">

As a user, I'd be happy to upload a key/value JSON dictionary, or  a command line path to docker even with additional labels. I have no good guidance on where to store user-specific data for this use case "as a web application". The only thing to manage is either a case insensitive map or to convert the addresses to checksum addresses. I looked at the /src/api/address-resolver/ impls and considered just adding to the hard-coded labels manually, but didn't think that was a good plan.

Thanks for considering/ listening ;)

**Comment by wmitsuda** [comment]
Yes, as a little hack, editing the files you pointed out would work.

I think specifying a user provided JSON file would be a good start, if it works for you, although not very user friendly. But I also don't want to block this feature just because I wasn't able to decide on a definitive solution for config persistence.

**Comment by totlsota** [comment]
I wasn't sure the best way to do that when running in docker?

**Comment by wmitsuda** [comment]
You can edit the `1.json` file at `hardcoded-addresses` directory, for example, and rebuild the docker image, where `1` is the chainid for mainnet. This hack should work for now.

But I agree that a proper solution is nice to have.

**Comment by sealer3** [comment]
As suggested, we could create a simple web application, maybe running on a different port, that handles the API for getting/setting private labels and stores them in a simple database. In the single-user case, it could be that easy, no authentication required. In the many-user case, we can use a large random hex string for authentication, like an "API key". In the public server case, need a spam prevention measure.

**Comment by JamesHinshelwood** [comment]
> The simplest way would be to store it in browser local storage, but I've heard some time ago Apple would start deleting local storage after some time.

I did a bit of research and I think the explanation is that [Safari deletes local storage after 7 days](https://webkit.org/blog/8613/intelligent-tracking-prevention-2-1/), unless the site is pinned to your home screen.

In my opinion, local storage is still the 'right' way to do this - it should work for most users. Maybe we can warn users that some browsers may delete local storage and they may want to export their data to somewhere else (e.g. via an "export to JSON" button).

Maybe its because I'm biased as a non-Safari user but local storage would be the cleanest and most convenient solution for me :stuck_out_tongue:

**Comment by wmitsuda** [comment]
yeah, I like the local storage solution bc it seems the natural choice for dapps.

I'm aware of the Safari policy, don't know how annoying it is in practice.
